coverage: Completely overhaul counter assignment, using node-flow graphs

Zalathar opened the pull request:
The existing code for choosing where to put physical counter-increments gets the job done, but is very ad-hoc and hard to modify without introducing tricky regressions.

This PR replaces all of that with a more principled approach, based on the algorithm described in "Optimal measurement points for program frequency counts" (Knuth & Stevenson, 1973).

---

We start by ensuring that our graph has “balanced flow”, i.e. each node's flow (execution count) is equal to the sum of all its in-edge flows, and equal to the sum of all its out-edge flows. That isn't naturally true of control-flow graphs, so we introduce a wrapper type `BalancedFlowGraph` to fix that by introducing synthetic nodes and edges as needed.

Once our graph has balanced flow, the next step is to create another view of that graph in which each node's successors have all been merged into one “supernode”. Consequently, each node's out-edges can be coalesced into a single out-edge to one of those supernodes. Because of the balanced-flow property, the flow of that coalesced edge is equal to the flow of the original node.

Having expressed all of our node flows as edge flows, we can then analyze node flows using techniques for analyzing edge flows. We incrementally build a spanning tree over the merged supernodes, such that each new edge in the spanning tree represents a node whose flow can be computed from that of other nodes.

When this is done, we end up with a list of “counter terms” for each node, describing which nodes need physical counters, and how the remaining nodes can have their flow calculated by adding and subtracting those physical counters.

---

The re-blessed coverage tests show that this results in modest or major improvements for our test programs. Some tests need fewer physical counters, some tests need fewer expression nodes for the same number of physical counters, and some tests show striking reductions in both.

Zalathar committed:
Add wrapper type `ReversedGraph` for swapping successors/predecessors

Zalathar committed:
Add `DenseBitSet::union_not`

This is similar to the existing `union`, except that bits in the RHS are
negated before being incorporated into the LHS.

Currently only `DenseBitSet` is supported. Supporting other bitset types is
possible, but non-trivial, and currently isn't needed.

Zalathar committed:
coverage: Completely overhaul counter assignment, using node-flow graphs

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12768381828/job/35588653999) [(plain)](https://github.com/rust-lang/rust/commit/25f4c560aeacafe0be09e771751ae3998dfa89d9/checks/35588653999/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error: unresolved link to `BalancedFlowGraph`
  --> compiler/rustc_mir_transform/src/coverage/counters/node_flow.rs:49:55
   |
49 |     /// though it does not necessarily have to be a [`BalancedFlowGraph`].
   |                                                       ^^^^^^^^^^^^^^^^^ no item named `BalancedFlowGraph` in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`


error: unresolved link to `Self::unify`
  --> compiler/rustc_mir_transform/src/coverage/counters/union_find.rs:24:21
   |
24 |     /// tall. See [`Self::unify`] for details.
   |                     ^^^^^^^^^^^ the struct `UnionFindEntry` has no field or associated item named `unify`
error: could not document `rustc_mir_transform`
warning: build failed, waiting for other jobs to finish...
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Tue Jan 14 13:39:31 UTC 2025
  network time: Tue, 14 Jan 2025 13:39:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Zalathar commented:
One of the nice things about this approach, compared to the status-quo, is that we only need to instrument control-flow nodes, not edges.

The current code instruments control-flow edges by inserting extra MIR blocks, and then adding counters to those blocks. That works, but it's nice to be able to only instrument nodes, so that the MIR structure isn't disturbed.

oli-obk reviewed (changes requested):
My third time reading the code in two days. Unfortunately I fail to build a full model of this change in my head. All the code is doing what the comments describe, and the code makes sense in small parts. I have absolutely no idea if it's doing the algorithm from the paper.

Uh... r=me with nits unless you want a reviewer that will actually grok the whole thing.

Zalathar commented on compiler/rustc_mir_transform/src/coverage/counters/node_flow.rs:
```diff
@@ -0,0 +1,290 @@
+
+        // Recursively make our immediate spantree-parent the root of what's
+        // left of its component, so that only one more edge rotation is needed.
+        self.yank_to_spantree_root(span_parent);
```
Hmm, the worst-case scenario is probably a long sequence of if-else statements, in which case we might have to traverse the whole thing.

I wonder if I can just make this iterative (with a recycled buffer) instead of recursive.

Zalathar committed:
coverage: Make `yank_to_spantree_root` iterative instead of recursive

This avoids having to worry about stack space when traversing very long
spantree paths, e.g. when instrumenting a long sequence of if/else statements.

Zalathar commented:
Added a FIXME, and made `yank_to_spantree_root` iterative [(diff)](https://github.com/rust-lang/rust/compare/d3fe1dd3f943573a177f710357f5d917b960abdf..9e9ca8e7b0cf39017534d2e39bb2fd3f4933f18f).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 9e9ca8e7b0cf39017534d2e39bb2fd3f4933f18f has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 9e9ca8e7b0cf39017534d2e39bb2fd3f4933f18f -->
<!-- homu: {"type":"Approved","sha":"9e9ca8e7b0cf39017534d2e39bb2fd3f4933f18f","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Ah whoops, just fixed a typo [(diff)](https://github.com/rust-lang/rust/compare/9e9ca8e7b0cf39017534d2e39bb2fd3f4933f18f..6eabf03526e561ea31b99e14a069e86cc168bede) and didn't see that this had been approved.

@bors r=oli-obk

bors commented:
:pushpin: Commit 6eabf03526e561ea31b99e14a069e86cc168bede has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 6eabf03526e561ea31b99e14a069e86cc168bede -->
<!-- homu: {"type":"Approved","sha":"6eabf03526e561ea31b99e14a069e86cc168bede","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
